File: firebase/firestore/lib/src/data/student.dart
import "package:meta/meta.dart";

import "package:firestore/constants.dart";
import "package:firestore/helpers.dart";

import "schedule.dart";

/// A user object.
/// 
/// At first, users need to be tracked by their IDs and personal data. After 
/// their schedules are compiled, those need to be added to the object. However,
/// since this class is [immutable], call [addSchedule] to receive a new 
/// [User] object with the added schedule data.
@immutable
class User extends Serializable {
	/// Warns for any users with no classes in their schedule.
	static void verifySchedules(Iterable<User> users) {
		final Set<User> missingSchedules = users.where(
			(User user) => user.hasNoClasses
		).toSet();

		if (missingSchedules.isNotEmpty) {
			// Warning since it can be a sign of data corruption.
			Logger.warning("Missing schedules for $missingSchedules");
		}
	}

	/// Converts a list of [Period] objects to JSON. 
	static List<Map<String, dynamic>> scheduleToJson(List<Period> schedule) => [
		for (final Period period in schedule) 
			period?.json
	];

	static final dayNamesList = List<String>.from(dayNames);

	/// This user's email.
	final String email;

	/// This user's first name.
	final String first;

	/// This user's last name.
	final String last;

	/// This user's ID.
	final String id;

	/// The section ID of this user's homeroom.
	final String homeroom;

	/// The location of this user's homeroom.
	final String homeroomLocation;

	/// This user's schedule.
	/// 
	/// This must have all the letters inside. 
	final Map<String, List<Period>> schedule;

	/// This user's full name.
	String get name => "$first $last";

	/// Creates a user object.
	User({
		@required this.first,
		@required this.last,
		@required this.email,
		@required this.id,
		this.homeroom,
		this.homeroomLocation,
		this.schedule,
	}) : 
		assert(id != null, "Could not find ID for user"),
		assert(
			first != null && last != null && email != null,
			"Could not find name for user: $id"
		)
	{
		if (schedule == null) {
			return;
		}
		assert(
			homeroom != null,
			"Could not find homeroom for user: ${toString()}"
		);
		assert(
			homeroomLocation != null,
			"Could not find homeroom location for user: ${toString()}"
		);
		for (final String dayName in dayNames) {
			assert(
				schedule.containsKey(dayName), 
				"$name does not have a schedule for $dayName days,"
			);
		}
	}

	/// If this user has no classes.
	bool get hasNoClasses => schedule.values.every(
		(List<Period> daySchedule) => daySchedule.every(
			(Period period) => period == null
		)
	);

	/// Returns a new [User] with added data. 
	/// 
	/// This fills in [homeroom] and [homeroomLocation], and is needed since 
	/// [User] objects are [immutable].
	User addHomeroom({
		@required String homeroom, 
		@required String homeroomLocation,
	}) => User(
		first: first,
		last: last,
		email: email,
		id: id,
		homeroom: homeroom,
		homeroomLocation: homeroomLocation,
	);

	/// Returns a new [User] with added data. 
	/// 
	/// This fills in [schedule], and is needed since [User] objects 
	/// are [immutable].
	/// 
	/// To fill in the homeroom as well, call this function on the return value
	/// of [addHomeroom].
	User addSchedule(Map<String, List<Period>> schedule) => User(
		first: first, 
		last: last,
		email: email,
		id: id,
		homeroom: homeroom,
		homeroomLocation: homeroomLocation,
		schedule: schedule,
	);

	@override
	Map<String, dynamic> get json => {
		for (final String dayName in dayNames) 
			dayName: scheduleToJson(schedule [dayName]),
		"homeroom": homeroom,
		"homeroom meeting room": homeroomLocation,
		"email": email,
		"dayNames": dayNamesList, 
	};

	@override
	String toString() => "$name ($id)";
}

File: firebase/firestore-py/lib/data/student.py
from .. import utils
from .schedule import Period

class DayDefaultDict(dict):
	def __missing__(self, letter):
		self[letter] = [None] * Period.PERIODS_IN_DAY[letter]
		return self[letter]
	def populate(self, keys): 
		for key in keys: 
			if key not in self: self.__missing__(key)

class User:
	def verify_schedule(users): 
		missing_schedules = {user for user in users if user.has_no_classes()}
		if (missing_schedules): 
			utils.logger.warning(f"Misisng schedules for {missing_schedules}")

	def schedule_to_json(schedule): return [
		period.to_json() if period is not None else None
		for period in schedule
	]

	def __init__(self, first, last, email, id, homeroom=None, homeroom_location=None, schedule=None):
		if not id:
			print(f"Could not find id for user: {first} {last}, {email}")
			print("Moving on")

		# assert first and last and email, f"Could not find contact info for {self}"
		self.first = first
		self.last = last
		self.name = f"{first} {last}"
		self.email = email
		self.id = id
		self.homeroom = homeroom
		self.homeroom_location = homeroom_location
		self.schedule = schedule
		if not schedule: return
		assert homeroom and homeroom_location, f"Could not find homeroom for {self}"
		for day_name in utils.constants.day_names: 
			assert day_name in schedule, f"{self} does not have a schedule for {day_name}"

	def empty(email, first, last): 
		user = User(
			first = first,
			last = last,
			email = email,
			id = "TEST",
			homeroom = "SENIOR_HOMEROOM",
			homeroom_location = "Unavailable",
			schedule = DayDefaultDict(),
		)
		user.schedule.populate(utils.constants.day_names)
		return user

	def __repr__(self):
		return f"{self.first} {self.last} ({self.id})"

	def has_no_classes(self):
		return all(
		all(period is None for period in day)
		for day in self.schedule
	)

	def to_json(self): return {
		**{
			day_name: User.schedule_to_json(self.schedule [day_name])
			for day_name in utils.constants.day_names
		},
		"advisory": None if not self.homeroom else {
			"id": self.homeroom, 
			"room": self.homeroom_location,
		},
		"email": self.email,
		"dayNames": utils.constants.day_names,
		"contactInfo": {
			"name": f"{self.first} {self.last}",
			"email": self.email
		}
	}
